src/graph/query_graph.c

```diff
@@ -93,6 +102,99 @@ static void _BuildQueryGraphAddEdge(QueryGraph *qg, const cypher_astnode_t *ast_
 	else QueryGraph_ConnectNodes(qg, dest, src, edge);
 }
 
+// Extracts node from 'qg' and places a copy of into 'graph'
+static void _QueryGraph_ExtractNode(const QueryGraph *qg, QueryGraph *graph,
+		AST *ast, const cypher_astnode_t *ast_node) {
+
+	// Validate inputs.
+	ASSERT(qg != NULL && graph != NULL && ast != NULL && ast_node != NULL);
+
+	// See if node is already in 'graph'.
+	const char *alias = AST_GetEntityName(ast, ast_node);
+	QGNode *n = QueryGraph_GetNodeByAlias(graph, alias);
+
+	if(n == NULL) {
+		// Node is missing from 'graph', try getting it from 'qg'.
+		n = QueryGraph_GetNodeByAlias(qg, alias);
+		if(n == NULL) {
+			// Node is missing from 'qg', create it.
+			_QueryGraphAddNode(graph, ast_node);
```

Comment:
you don't want to update `THE query graph`?

Reply:
No, consider:
`MATCH (a) WITH a WHERE (a)-[]->(b)`
There are 2 scopes here:
1.`MATCH (a) WITH a`
2.`WHERE (a)-[]->(b)`

`qg` belongs to the first scope and contains only `a`.
we shouldn't extend it.